src/utils/cache.h

```diff
@@ -322,4 +328,293 @@ class LruCacheLock {
   V* value_ = nullptr;
 };
 
+// A hash-keyed cache. Thread-safe. Takes ownership of all values, which are
+// deleted upon eviction; thus, using values stored requires pinning them, which
+// in turn requires Unpin()ing them after use. The use of HashKeyedCacheLock is
+// recommend to automate this element-memory management.
+// Unlike LRUCache, doesn't even consider trying to support LRU order.
+// Does not support delete.
+// Does not support replace! Inserts to existing elements are silently ignored.
+// FIFO eviction.
+// Assumes that eviction while pinned is rare enough to not need to optimize unpin for that case.
+template <class V>
```

Comment:
It makes sense to put it into a separate `.h` file, or rename existing one to `cache-old-v2.h` like we did with https://github.com/LeelaChessZero/lc0/blob/master/src/utils/cache-old.h and keep this in cache.h.

Reply:
I've dropped this from 1490 for now - I'll revisit it in a future PR.